### PR TITLE
Unambiguous ownership of generic param objects

### PR DESCRIPTION
**Pull request by xmcclure**
This patch adds a small number of mempool-reference audit points to functions in class.c and then performs a large refactor without which those audit points would fail on the standard tests. The goal of the refactor is to remove ambiguities in what image owns or should own any particular generic parameter object. Previously, there were four options for specifying a MonoGenericParam’s image:
	1. klass/method union in MonoGenericContainer.
	2. image field in MonoGenericContainer.
	3. image field in MonoGenericParam.
	4. No field; “image” passed in to methods in at time of usage.
This patch removes 3 and 4; it preserves the image field from (2), but merges it into the union from (1). It adds a new field, “is_anonymous”, to indicate when the image member of the union should be used. The single source of the “owner" union now always points to the generic param’s image, directly or indirectly.

A further change is made to support this: Previously, containers were optional, and the presence or absence of container was incidentally used to determine whether an object was a MonoGenericParam or a MonoGenericParamFull. Now, containers are mandatory for all MonoGenericParams, so a new “is_small_size” bit in the container tracks this information.

Some changes were made to checked builds to support all this: there's a g_assert_checked now which asserts but only on checked builds, and a bunch of little changes to the mempool audit feature (see commit #1)

This change opens some sources of risk:
	⁃ Code which creates MonoGenericParams now must make sure to allocate a container early.
	⁃ Code that accesses the MonoGenericContainer::owner field must all be correctly updated to look for is_anonymous or there will be a crash. (Much of this code would have previously used a NULL container as the flag for “anonymous”).
	⁃ Code that casts MonoGenericParams to MonoGenericParamFull must use the is_small_size field.
	⁃ Some generic containers will now exist with NULL type_params and 0 type_argc. Code which reads the type param array must either not receive objects from these containers, or must handle it.
	⁃ If code by users of the embedding API exists which somehow depends on particular behavior related to the now-unused image argument of mono_class_from_generic_parameter, it will break.
I have audited code around, at least, the first three of these cases and fixed what problems I found.

Monodis is probably completely broken now.

### CODE REVIEW

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -2507,10 +2507,9 @@ encode_klass_ref_inner (MonoAotCompile *acfg, MonoClass *klass, guint8 *buf, gui
 		encode_value (klass->byval_arg.type, p, &p);
 		encode_value (mono_type_get_generic_param_num (&klass->byval_arg), p, &p);
 
-		encode_value (container ? 1 : 0, p, &p);
-		if (container) {
+		encode_value (container->is_anonymous ? 0 : 1, p, &p);
+		if (!container->is_anonymous) {
 			encode_value (container->is_method, p, &p);
-			g_assert (!par->gshared_constraint);
```
Comment (xmcclure) [inline mono/mini/aot-compiler.c]: This was removed under Zoltan's supervision but I'm actually sort of confused-- why was this removed? Is this okay to remove? With all the other changes the assert now fails.

**Comment by kumpera** [comment]
Avoid merge commits in PR as much as possible, most git tools suck at presenting those merges in any usable way, making offline reviews a lot harder as it's no longer the case that the PR is made out of the top X commits.

File: mono/metadata/image.c [hunk 2]
```diff
@@ -2538,49 +2538,24 @@ mono_image_append_class_to_reflection_info_set (MonoClass *klass)
...
-	return mono_mempool_no_owner;
-}
+// This is support for the mempool reference tracking feature in checked-build, but lives in image.c due to use of static variables of this file.
 
 /**
- * mono_find_mempool_owner:
+ * mono_find_image_owner:
  *
- * Find the image or imageset, if any, which a given pointer is located in the memory of.
+ * Find the disk-loaded image, if any, which a given pointer is located in the memory of.
```
Comment (kumpera) [inline mono/metadata/image.c]: mages don't need to be disk-loaded, there are other ways to load them besides disk files that will in turn produce non-dynamic images. mkbundle'd assemblies and Assembly.Load (byte[]) are two examples of that.

File: mono/metadata/image.c [hunk 3]
```diff
@@ -2588,14 +2563,12 @@ mono_find_mempool_owner (void *ptr)
 
 		// Iterate over images within a hash
 		g_hash_table_iter_init (&iter, target);
-		while (searching && g_hash_table_iter_next(&iter, NULL, (gpointer *)&image))
+		while (!owner && g_hash_table_iter_next(&iter, NULL, (gpointer *)&image))
 		{
 			mono_image_lock (image);
```
Comment (kumpera) [inline mono/metadata/image.c]: I just had the sad realization that this is nesting images_lock with image_lock, which is not part of the regular locking rules.

In addition to that, we're taking those two locks at the audit points, which might be violating other locking rules.
A quick check of currents uses point to this being currently ok.

File: mono/metadata/metadata.c [hunk 4]
```diff
@@ -6638,3 +6638,36 @@ mono_method_get_wrapper_cache (MonoMethod *method)
...
+ *
+ * Find the imageset, if any, which a given pointer is located in the memory of.
+ */
+MonoImageSet *
+mono_find_image_set_owner (void *ptr)
+{
+	MonoImageSet *owner = NULL;
+	int i;
+
+	image_sets_lock ();
```
Comment (kumpera) [inline mono/metadata/metadata.c]: The same concern about locking applies here.

File: mono/metadata/metadata.c [hunk 5]
```diff
@@ -6638,3 +6638,36 @@ mono_method_get_wrapper_cache (MonoMethod *method)
 		return &method->klass->image->wrapper_caches;
 	}
 }
+
+#if CHECKED_BUILD
```
Comment (kumpera) [inline mono/metadata/metadata.c]: We should try to avoid this sort of conditional compilation creep.

It makes for a more fragile code base, say I break it tomorrow, we won't pick that
until that until the next time we do a checked build.

In addition to that, now it's one extra configuration to keep in mind when casually looking at that code.

The nice thing is that we don't actually need to worry about possible extra bloat from this function because it has non-public visibility and the linker will remove it.

File: mono/metadata/reflection.c [hunk 6]
```diff
@@ -215,6 +215,54 @@ static void init_type_builder_generics (MonoObject *type);
 #define ADDP_IS_GREATER_OR_OVF(a, b, c) (((a) + (b) > (c)) || CHECK_ADDP_OVERFLOW_UN (a, b))
 #define ADD_IS_GREATER_OR_OVF(a, b, c) (((a) + (b) > (c)) || CHECK_ADD4_OVERFLOW_UN (a, b))
 
+#if CHECKED_BUILD
```
Comment (kumpera) [inline mono/metadata/reflection.c]: I don't think we should make this conditional on CHECKED_BUILD.

**Comment by kumpera** [comment]
The first commit, 2bcb4f001309feed42447f04b340df8018f15999, needs some minor edits, but LGTM otherwise.

File: mono/metadata/class-internals.h [hunk 7]
```diff
@@ -609,22 +602,23 @@ struct _MonoGenericContainer {
...
-	int type_argc    : 31;
+	int type_argc    : 29; // Per the ECMA spec, this value is capped at 16 bits
 	/* If true, we're a generic method, otherwise a generic type definition. */
 	/* Invariant: parent != NULL => is_method */
-	int is_method    : 1;
+	int is_method     : 1;
+	/* If true, this is one of the special anonymous containers in MonoImage, or a container created by SRE. */
+	int is_anonymous : 1;
+	/* If false, all params in this container are full-size. If true, all params are just param structs. */
+	/* This field is always == to the is_anonymous field except during container creation. */
```
Comment (kumpera) [inline mono/metadata/class-internals.h]: "except during container creation" Why? For how long? How will code witness this transitory state?

Please expand this comment on what's the issue around container creation.
Doing it while the knowledge is fresh will save us from a lot of trouble down the line.

File: mono/metadata/class-internals.h [hunk 8]
```diff
@@ -632,8 +626,9 @@ struct _MonoGenericContainer {
 
 #define mono_generic_param_owner(p)		((p)->owner)
 #define mono_generic_param_num(p)		((p)->num)
-#define mono_generic_param_info(p)		(mono_generic_param_owner (p) ? &((MonoGenericParamFull *) p)->info : NULL)
-#define mono_generic_param_name(p)		(((mono_generic_param_owner (p) || (p)->gshared_constraint)) ? ((MonoGenericParamFull *) p)->info.name : NULL)
+#define mono_generic_param_is_fullsize(p) ( !mono_generic_param_owner (p)->is_small_param )
```
Comment (kumpera) [inline mono/metadata/class-internals.h]: I'm with Zoltan on this one.

Those macros are now overly complicated and deserve to be turned into static inline functions.

File: mono/metadata/class.c [hunk 9]
```diff
@@ -4268,7 +4269,7 @@ verify_class_overrides (MonoClass *klass, MonoMethod **overrides, int onum)
 		}
 
 		if (!mono_class_is_assignable_from_slow (decl->klass, klass)) {
-			mono_class_set_failure (klass, MONO_EXCEPTION_TYPE_LOAD, g_strdup ("Method overrides a class or interface that extended or implemented by this type"));
+			mono_class_set_failure (klass, MONO_EXCEPTION_TYPE_LOAD, g_strdup ("Method overrides a class or interface that is not extended or implemented by this type"));
```
Comment (kumpera) [inline mono/metadata/class.c]: Unrelated, it should in outside of this PR.

File: mono/metadata/class-internals.h [hunk 10]
```diff
@@ -66,7 +66,7 @@ struct _MonoMethod {
 	guint16 flags;  /* method flags */
 	guint16 iflags; /* method implementation flags */
 	guint32 token;
-	MonoClass *klass;
+	MonoClass *klass; /* To what class does this method belong */
```
Comment (kumpera) [inline mono/metadata/class-internals.h]: Unrelated, it should go in outside of this PR/commit.

File: mono/metadata/class.c [hunk 11]
```diff
@@ -5668,7 +5669,7 @@ mono_class_setup_parent (MonoClass *klass, MonoClass *parent)
  *    hierarchy + 1
  *  - supertypes: array of classes: each element has a class in the hierarchy
  *    starting from @class up to System.Object
- * 
+ *
```
Comment (kumpera) [inline mono/metadata/class.c]: Line damage

File: mono/metadata/class.c [hunk 12]
```diff
@@ -5692,13 +5693,21 @@ mono_class_setup_supertypes (MonoClass *klass)
 	supertypes = mono_class_alloc0 (klass, sizeof (MonoClass *) * ms);
 
 	if (klass->parent) {
-		supertypes [klass->idepth - 1] = klass;
+		CHECKED_METADATA_WRITE_PTR ( supertypes [klass->idepth - 1] , klass );
+
+		// Reference checking is done one assignment at a time, so memcpy cannot be used in checked mode
+#ifdef CHECKED_BUILD
```
Comment (kumpera) [inline mono/metadata/class.c]: We should shift as much complexity as possible to checked build itself.

This one-off configuration changes eventually pile up and make code too hard to maintain.

Checked build should not introduce additional complication into the runtime code in the form of extra conditionals.

File: mono/metadata/class.c [hunk 13]
```diff
@@ -5808,6 +5817,7 @@ mono_class_create_from_typedef (MonoImage *image, guint32 type_token, MonoError
 	if (klass->generic_container) {
 		klass->is_generic = 1;
 		klass->generic_container->owner.klass = klass;
+		klass->generic_container->is_anonymous = FALSE; // Owner class is now known, container is no longer anonymous
```
Comment (kumpera) [inline mono/metadata/class.c]: This looks like a very odd place to initialize is_anonimous.

Couldn't it be done in mono_metadata_load_generic_params?

File: mono/metadata/class.c [hunk 14]
```diff
@@ -5986,7 +5996,7 @@ mono_class_is_nullable (MonoClass *klass)
 }
 
 
-/** if klass is T? return T */
+/** For a particular Nullable<T>, get the class of the T */
```
Comment (kumpera) [inline mono/metadata/class.c]: This is saying the same thing. T? is the C# form for Nullable<T>.

File: mono/metadata/class.c [hunk 15]
```diff
@@ -6070,6 +6080,7 @@ mono_generic_class_get_class (MonoGenericClass *gclass)
 
 	klass->cast_class = klass->element_class = klass;
 
+	// Nullable has special rules when it comes to casting and boxing
```
Comment (kumpera) [inline mono/metadata/class.c]: This is a bad comment, it creates more confusion than it answers.

File: mono/metadata/class.c [hunk 16]
```diff
@@ -6117,42 +6128,84 @@ mono_generic_class_get_class (MonoGenericClass *gclass)
...
+	MonoImage *image = get_image_for_container (container);
+	gboolean is_mvar = container->is_method;
+	gboolean is_anonymous = container->is_anonymous;
 
 	klass = mono_image_alloc0 (image, sizeof (MonoClass));
 	classes_size += sizeof (MonoClass);
 
 	if (pinfo) {
-		klass->name = pinfo->name;
+		CHECKED_METADATA_WRITE_PTR_EXEMPT ( klass->name, pinfo->name );
```
Comment (kumpera) [inline mono/metadata/class.c]: Mixing adding the checks with the core change makes reviewing harder.

It makes for worse commit history as it make the same commit show up in more places than needed   when checking log or blame for a file.

File: mono/metadata/class.c [hunk 17]
```diff
@@ -6312,66 +6375,70 @@ set_anon_gparam_class (MonoGenericParam *param, gboolean is_mvar, MonoClass *kla
...
 MonoClass *
-mono_class_from_generic_parameter (MonoGenericParam *param, MonoImage *image, gboolean is_mvar)
+mono_class_from_generic_parameter_internal (MonoGenericParam *param)
 {
-	MonoGenericContainer *container = mono_generic_param_owner (param);
-	MonoGenericParamInfo *pinfo = NULL;
-	MonoClass *klass, *klass2;
+	MonoImage *image = get_image_for_generic_param (param);
+	MonoGenericParamInfo *pinfo = mono_generic_param_info (param);
+	MonoClass *klass, *klass_other_thread;
```
Comment (kumpera) [inline mono/metadata/class.c]: In general, we don't rename variables unless they have the wrong name, it destroys history more than helping the code.

In this case is ok as all uses had to be changed anyways.

File: mono/metadata/metadata.c [hunk 18]
```diff
@@ -2351,49 +2320,58 @@ image_sets_unlock (void)
 static MonoImageSet*
 get_image_set (MonoImage **images, int nimages)
 {
+	/* Cache used only in this function. Protected by image_sets_mutex */
+	static MonoImageSet *mscorlib_image_set;
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Given this function is not part of this change, I think it's best to leave it completely unchanged.

File: mono/metadata/metadata.c [hunk 19]
```diff
@@ -2438,7 +2420,8 @@ delete_image_set (MonoImageSet *set)
 	for (i = 0; i < set->nimages; ++i)
 		set->images [i]->image_sets = g_slist_remove (set->images [i]->image_sets, set);
 
-	g_ptr_array_remove (image_sets, set);
+	if (image_sets)
```
Comment (kumpera) [inline mono/metadata/metadata.c]: This can't happen, the list must be created for an image set to exist.

Do you have an example of this not being true?

File: mono/metadata/metadata.c [hunk 20]
```diff
@@ -2404,10 +2382,14 @@ get_image_set (MonoImage **images, int nimages)
 		set->ginst_cache = g_hash_table_new_full (mono_metadata_generic_inst_hash, mono_metadata_generic_inst_equal, NULL, (GDestroyNotify)free_generic_inst);
 		set->gmethod_cache = g_hash_table_new_full (inflated_method_hash, inflated_method_equal, NULL, (GDestroyNotify)free_inflated_method);
 		set->gsignature_cache = g_hash_table_new_full (inflated_signature_hash, inflated_signature_equal, NULL, (GDestroyNotify)free_inflated_signature);
-	
+
 		for (i = 0; i < nimages; ++i)
 			set->images [i]->image_sets = g_slist_prepend (set->images [i]->image_sets, set);
 
+		// Add to global image sets list
+		if (!image_sets)
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Why move initialization?

File: mono/metadata/metadata.c [hunk 21]
```diff
@@ -2637,48 +2638,13 @@ collect_type_images (MonoType *type, CollectData *data)
...
-			else
-				add_image (container->owner.klass->image, data);
-		} else {
-			add_image (type->data.generic_param->image, data);
-		}
-	}
+	case MONO_TYPE_VAR:
+	case MONO_TYPE_MVAR:
+	{
+		MonoImage *image = get_image_for_generic_param (type->data.generic_param);
```
Comment (kumpera) [inline mono/metadata/metadata.c]: How about the refemit comment on the MVAR case?

File: mono/metadata/metadata.c [hunk 22]
```diff
@@ -2944,7 +2910,7 @@ mono_metadata_get_image_set_for_method (MonoMethodInflated *method)
 MonoGenericInst *
 mono_metadata_get_generic_inst (int type_argc, MonoType **type_argv)
 {
-	MonoGenericInst *ginst;
+	MonoGenericInst *ginst, *ginst_key;
```
Comment (kumpera) [inline mono/metadata/metadata.c]: This name change should not go in. It's not worth to have it.

File: mono/metadata/metadata.c [hunk 23]
```diff
@@ -3016,12 +2982,11 @@ mono_metadata_lookup_generic_class (MonoClass *container_class, MonoGenericInst
 	MonoImageSet *set;
 	CollectData data;
 
+	memset (&helper, 0, sizeof(helper)); // act like g_new0
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Is this actually changing anything? If not, just take it out.

File: mono/metadata/metadata.c [hunk 24]
```diff
@@ -3195,7 +3160,29 @@ select_container (MonoGenericContainer *gc, MonoTypeEnum type)
...
+MonoGenericContainer *
+get_anonymous_container_for_image (MonoImage *image, gboolean is_mvar)
+{
+	MonoGenericContainer **container_pointer;
+	if (is_mvar)
+		container_pointer = &image->anonymous_generic_method_container;
+	else
+		container_pointer = &image->anonymous_generic_class_container;
+	// This container has never been created; make it now.
+	if (!*container_pointer)
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Use a temporary here and avoid all of this indirection.

File: mono/metadata/metadata.c [hunk 25]
```diff
@@ -3195,7 +3160,29 @@ select_container (MonoGenericContainer *gc, MonoTypeEnum type)
...
+	{
+		// Note this is never deallocated anywhere-- it exists for the lifetime of the image it's allocated from
+		*container_pointer = mono_image_alloc0 (image, sizeof (MonoGenericContainer));
```
Comment (kumpera) [inline mono/metadata/metadata.c]: This is publishing uninitialized data as there's no locking around the parsing code.
It might cause other threads to see partially initialized data.

The right way is to init on a temporary and only at the end store it to the right place.

Then the question is whether the anon container must be unique or not.

If it doesn't need to be unique, a store_release is enough (and it might cause duplicates).

If it does, CAS the value in.

File: mono/metadata/metadata.c [hunk 26]
```diff
@@ -3214,12 +3201,24 @@ mono_metadata_parse_generic_param (MonoImage *m, MonoGenericContainer *generic_c
...
+		gboolean is_mvar = FALSE;
+		switch (type)
+		{
+			case MONO_TYPE_VAR:
+				break;
+			case MONO_TYPE_MVAR:
+				is_mvar = TRUE;
+				break;
+			default:
+				g_assert (FALSE); // This is not a generic param and we should not be creating a generic param object.
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Use g_error with a message.

File: mono/metadata/metadata.c [hunk 27]
```diff
@@ -4935,8 +4934,7 @@ mono_metadata_generic_param_equal_internal (MonoGenericParam *p1, MonoGenericPar
 	 * The AOT runtime doesn't set the image when it's decoding
 	 * types, so we only compare it when the owner is NULL.
 	 */
-	if (mono_generic_param_owner (p1) == mono_generic_param_owner (p2) &&
-	    (mono_generic_param_owner (p1) || p1->image == p2->image))
+	if (mono_generic_param_owner (p1) == mono_generic_param_owner (p2))
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Looks like the comment above this conditional is now stale as we're no longer checking images.

File: mono/metadata/metadata.c [hunk 28]
```diff
@@ -5014,6 +5012,7 @@ mono_metadata_fnptr_equal (MonoMethodSignature *s1, MonoMethodSignature *s2, gbo
  * mono_metadata_type_equal:
  * @t1: a type
  * @t2: another type
+ * @signature_only: If true, treat generics which are instantiated separately but "shaped the same" as equal
```
Comment (kumpera) [inline mono/metadata/metadata.c]: Shaped the same is ambiguous. It's simple, they are treated just for their positional values instead of container + position.

File: mono/mini/aot-runtime.c [hunk 29]
```diff
@@ -471,18 +472,16 @@ decode_klass_ref (MonoAotModule *module, guint8 *buf, guint8 **endbuf)
...
+		gboolean is_not_anonymous = decode_value (p, &p);
 		MonoType *gshared_constraint = NULL;
-		char *par_name = NULL;
 
-		t = NULL;
-		if (has_container) {
+		if (is_not_anonymous) {
 			gboolean is_method = decode_value (p, &p);
-			
+
```
Comment (kumpera) [inline mono/mini/aot-runtime.c]: line damage

File: mono/mini/aot-runtime.c [hunk 30]
```diff
@@ -515,6 +514,9 @@ decode_klass_ref (MonoAotModule *module, guint8 *buf, guint8 **endbuf)
 
 				t = mini_get_shared_gparam (&par_klass->byval_arg, gshared_constraint);
 			}
+
+            // FIXME: We didn't decode is_method, so this arg is always FALSE. Is this a problem?
```
Comment (kumpera) [inline mono/mini/aot-runtime.c]: Yes, it will be a problem since we won't be able to create anon MVARs.

File: mono/metadata/image.c (hunk 3 above)
Comment (xmcclure) [inline mono/metadata/image.c]: Should I do something about this?

The audits are really not going to be possible if we can't acquire locks. As for nesting image_lock inside images_lock, the only way I can think to get around that is to make a copy of the list of all images in the system… every time… which will be a bit performance painful maybe even by the checked build standards.

File: mono/metadata/class.c (hunk 13 above)
Comment (xmcclure) [inline mono/metadata/class.c]: I'm extending the comments on `is_anonymous` to clarify, but: `mono_metadata_load_generic_params` cannot set the owner because it does not know who the owner is, the owner is always assigned by `mono_metadata_load_generic_params`'s caller. Until the owner is known, `is_anonymous` must be TRUE. In other words, `is_anonymous` is not "initialized" here; it changes.

It would be possible for `mono_metadata_load_generic_params` to take some representation of the owner as an argument. If we did this, we would remove the quirk where some containers are "temporarily" anonymous during initialization. Would you prefer this approach?

File: mono/metadata/class.c (hunk 14 above)
Comment (xmcclure) [inline mono/metadata/class.c]: Honestly the original wording confused me because I wasn't sure if T? was a unicode error ^_^; but I can change it back.

File: mono/metadata/class.c (hunk 17 above)
Comment (xmcclure) [inline mono/metadata/class.c]: I went ahead and backed it out and added a comment, a lot of the klass2 lines have no other changes.

File: mono/metadata/metadata.c (hunk 19 above)
Comment (xmcclure) [inline mono/metadata/metadata.c]: I was thinking that technically you could call this function out of order with an improperly created image set, but the function isn't visible outside this file and all uses in the file are correct so I guess that isn't a real concern.

File: mono/metadata/metadata.c (hunk 21 above)
Comment (xmcclure) [inline mono/metadata/metadata.c]: Discussed this with Rodrigo out of band

File: mono/metadata/metadata.c (hunk 23 above)
Comment (xmcclure) [inline mono/metadata/metadata.c]: I do think this should stay-- there are two problems here:

1. This code sets NULL all fields needed by `collect_gclass_images`, but this is essentially a coincidence. Someone who edited `collect_gclass_images` would not know that they could cause a crash by using new fields-- the contract that the function only uses those fields are used is totally implicit.

2. Passing around partially-invalid objects complicates debugging. If I remember correctly this one came up for me because I was looking at the object after it had escaped this function, I tried looking at the owner so I could try to get a name, owner was set to a non-NULL object and I wound up getting gibberish. This lead to some wasted time trying to figure out if there was heap corruption before realizing I had just read an uninitialized pointer, since we're so consistent about initializing entire structures to NULL elsewhere…

I think initializing to zero is consistent with other uses of this key-object-on-the-stack.

File: mono/metadata/metadata.c (hunk 27 above)
Comment (xmcclure) [inline mono/metadata/metadata.c]: I removed the second paragraph but left the first one.

File: mono/metadata/class.c (hunk 13 above)
Comment (kumpera) [inline mono/metadata/class.c]: Now it makes sense. Yeah, it would be nice if the container state is not changed after it's created.

File: mono/metadata/class.c (hunk 12 above)
Comment (xmcclure) [inline mono/metadata/class.c]: This is going to have to mean the memcpy goes away completely.

**Comment by xmcclure** [comment]
Rebased, addressed all comments above except those I specifically responded to, split out the changes into four clearly delineated commits instead of two.

**Comment by xmcclure** [comment]
Converted those macros to static inlines as Zoltan suggested

**Comment by xmcclure** [comment]
builddeb

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=30
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-conc-noturbo-binary/builds/2
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=31
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo-binary/builds/3
`</botmode>`

**Comment by kumpera** [comment]
Uggg, I had to manually revert this PR as it broke corlib compilation is some weird way I could not easily debug. Sorry. :(
